services/src/main/java/org/keycloak/services/managers/DefaultBruteForceProtector.java

```diff
@@ -168,20 +172,15 @@ protected void failure(KeycloakSession session, LoginEvent event) {
             userLoginFailure.incrementFailures();
             logger.debugv("new num failures: {0}", userLoginFailure.getNumFailures());
 
-            if(userLoginFailure.getNumFailures() == realm.getFailureFactor()) {
+            if(userLoginFailure.getNumFailures() >= realm.getFailureFactor()) {
```

Comment:
I noticed a tiny bug, which I fixed here: 

when changing brute force protector settings meanwhile brute force protector was already active, it could happen that number of failures got greater than failure factor and `==` was never true.

Reply:
Good point. Still I'm wondering if this condition might now need addition clauses as we want it to fire only once per lockout. So one might need to check that the user is either enabled or the disabled reason is not set to `DISABLED_BY_PERMANENT_LOCKOUT`. I have to admit I'm only looking at the local code and not at testing this in my IDE.

Reply:
I will test this under debugger.

Reply:
It seems will not be another call anymore to `DefaultBruteForceProtector.failure()` method after the user permanently is locked, and therefore the condition is never checked again. 

The processing of login is stopped already earlier, during the authentication flows when validating the user, for example here

https://github.com/keycloak/keycloak/blob/50afa72266a8a1c8b20413912564718f096c4209/services/src/main/java/org/keycloak/authentication/authenticators/browser/AbstractUsernameFormAuthenticator.java#L228

or here

https://github.com/keycloak/keycloak/blob/50afa72266a8a1c8b20413912564718f096c4209/services/src/main/java/org/keycloak/authentication/authenticators/directgrant/ValidateUsername.java#L80

Reply:
Even though the code does not seem to be called, if you think it would be safer I could still add a new condition, for example like this

```diff
@@ -174,7 +173,7 @@ public class DefaultBruteForceProtector implements Runnable, BruteForceProtector

             if(userLoginFailure.getNumFailures() >= realm.getFailureFactor()) {
                 UserModel user = session.users().getUserById(realm, userId);
-                if (user == null) {
+                if (user == null || user.isEnabled() == false) {
                     return;
                 }
                 logger.debugv("user {0} locked permanently due to too many login attempts", user.getUsername());
```

Reply:
Does it make more sense to be consistent with the original scope of this PR and avoid fixing the brute-force logic?

Reply:
Sure, I removed the fix for the comparison of number of failures.